### PR TITLE
support for adding db specific types

### PR DESCRIPTION
**Pull request by vikask-gs**
This has been refactored so that we can integrate it with db extension, in future, and is in line with what we agreed to earlier. Good to merge.

### CODE REVIEW

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/executionPlan/tests/executionPlanTest.pure [hunk 1]
```diff
@@ -1,1699 +1,1714 @@
...
+}
+
+function <<test.Test>> meta::pure::executionPlan::tests::testSupportStreamFlagWithTdsJoinForOneDB():Boolean[1]
+{
+   let plan =executionPlan({name:String[*]|
+                testJoinTDS_Person.all()->filter(p|$p.firstName->in($name))->meta::pure::tds::project([col(p|$p.firstName, 'firstName'), col(p|$p.employerID, 'eID')])
+                ->join(testJoinTDS_Firm.all()->filter(p|$p.legalName->in($name))->project([col(p|$p.firmID, 'fID'), col(p|$p.legalName, 'legalName')]), JoinType.INNER, {a,b|$a.getInteger('eID') == $b.getInteger('fID');});},testJoinTDSMapping, meta::relational::tests::testRuntime(), meta::pure::router::extension::defaultRelationalExtensions());
+   let functionParameters = $plan.rootExecutionNode->allNodes(defaultRelationalExtensions())->filter(node | $node->instanceOf(FunctionParametersValidationNode))->cast(@FunctionParametersValidationNode).functionParameters;
+
+   assertEquals($functionParameters->filter(param | $param.name == 'name').supportsStream, true)
```
Comment (epsstan) [inline legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/executionPlan/tests/executionPlanTest.pure:1714]: This is such a big diff :(

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/relationalTypeTranslations.pure [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+function meta::relational::translation::translateCoreTypeToDbSpecificType(type: meta::relational::metamodel::datatype::DataType[0..1], translationContext : TranslationContext[1]): meta::relational::metamodel::datatype::DataType[0..1]
+{
+   if($type->isEmpty() || $translationContext.dbType->isEmpty(), |$type,
+      |let translatedType = translateCoreTypeToDbSpecificTypeMap($type->toOne())->get($translationContext.dbType->toOne());
+      if($translatedType->isEmpty(), |$type, |$translatedType););
+}
+
+function <<access.private>> meta::relational::translation::translateCoreTypeToDbSpecificTypeMap(type: meta::relational::metamodel::datatype::DataType[1]):Map<DatabaseType,meta::relational::metamodel::datatype::DataType>[1]
+{
```
Comment (epsstan) [inline legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/relationalTypeTranslations.pure:19]: This function has db specific logic. Shouldn't db specific logic be in db specific abstractions. 

See my comment in #465

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/tests/testRelationalExtension.pure [hunk 3]
```diff
@@ -30,6 +31,12 @@ function <<test.Test>> meta::relational::metamodel::execute::tests::testExecuteI
    true;
 }
 
+function <<test.Test>> meta::relational::tests::typeInference::testTranslateDbType():Boolean[1]
```
Comment (epsstan) [inline legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/tests/testRelationalExtension.pure:34]: Can we add a comment/TODO to remind us that we should move this into the new db specific test structure

File: legend-pure-m2-store-relational/src/main/resources/platform/relational/relational.pure [hunk 4]
```diff
@@ -390,97 +390,114 @@ Class meta::relational::metamodel::RelationalOperationElementWithJoin extends me
...
 }
 
-Class meta::relational::metamodel::datatype::Array extends meta::relational::metamodel::datatype::DataType
+Class meta::relational::metamodel::datatype::Array extends meta::relational::metamodel::datatype::CoreDataType
 {
 }
 
+// Db Specific DataTypes
+// Big Query DataTypes
+Class meta::relational::metamodel::datatype::BigQueryString extends meta::relational::metamodel::datatype::DbSpecificDataType
```
Comment (epsstan) [inline legend-pure-m2-store-relational/src/main/resources/platform/relational/relational.pure:488]: What about the numeric BigQuery types ? https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#numeric_types

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/executionPlan/tests/executionPlanTest.pure (hunk 1 above)
Comment (vikask-gs) [reply legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/executionPlan/tests/executionPlanTest.pure:1714]: there might be a spaces/newline related issue. Please configure your diff viewer to ignore them

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/relationalTypeTranslations.pure (hunk 2 above)
Comment (vikask-gs) [reply legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/relationalTypeTranslations.pure:19]: yes, this is being discussed with pierre/david, on how should we refactor this, as this stuff is spread across platform and core_relational

File: legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/tests/testRelationalExtension.pure (hunk 3 above)
Comment (vikask-gs) [reply legend-pure-code-compiled-core-relational/src/main/resources/core_relational/relational/tests/testRelationalExtension.pure:34]: Sure

File: legend-pure-m2-store-relational/src/main/resources/platform/relational/relational.pure (hunk 4 above)
Comment (vikask-gs) [reply legend-pure-m2-store-relational/src/main/resources/platform/relational/relational.pure:488]: we were thinking of adding other types once the basic framework is in

**Review by davidharte-gs — APPROVED** [review]
Approved

**Pull request closed without merging**